HDDS-3217. Datanode startup is slow due to iterating container DB 2-3 times.

bharatviswa504 opened the pull request:
## What changes were proposed in this pull request?

Reduce the time spent during DN restart. Commit the required info which is needed during DN restart commit it to container DB.

## What is the link to the Apache JIRA

https://issues.apache.org/jira/browse/HDDS-3217

## How was this patch tested?

Updated Block Deletion test, will check if any more tests can be added.

bharatviswa504 commented:
Thank You @bshashikant for the review.
Added code changes to handle upgrade. This is brought during an offline discussion with @hanishakoneru .

@hanishakoneru Handled upgrade scenario also.

bharatviswa504 commented:
Last run of acceptance test failed due to DISK_OUT_OF_SPACE.
Kick-off test again.
`datanode_1  | 2020-04-17 23:17:31,752 [ChunkWriter-10-0] INFO keyvalue.KeyValueHandler: Operation: CreateContainer , Trace ID: 27ad02faf5f8ae77:9cf5a0e67938f85:27ad02faf5f8ae77:0 , Message: Container creation failed, due to disk out of space , Result: DISK_OUT_OF_SPACE , StorageContainerException Occurred`.

hanishakoneru commented:
Thank you Bharat for working on this. I have one suggestion. 
ChunkLayoutVersion is always gonna have FILE_PER_BLOCK and FILE_PER_CHUNK. Anytime we need to make a change i.e. add a new LayoutVersion, we will end up having to add 2 LayoutVersions to take into account these 2 types of chunk storage. Instead, we can we have a separate LayoutVersion for non-chunk related versioning. Current changes could probably go under BlockLayoutVersion. What do you think?

bharatviswa504 commented:
> Thank you Bharat for working on this. I have one suggestion.
> ChunkLayoutVersion is always gonna have FILE_PER_BLOCK and FILE_PER_CHUNK. Anytime we need to make a change i.e. add a new LayoutVersion, we will end up having to add 2 LayoutVersions to take into account these 2 types of chunk storage. Instead, we can we have a separate LayoutVersion for non-chunk related versioning. Current changes could probably go under BlockLayoutVersion. What do you think?

Thanks for the review. I like the idea, when implementing also i thought to do this i need to add one more layout version field to the container file. And Old Containers will not have this field at all. Instead of doing all these, I have taken a simple approach of reusing chunkLayoutVersion. And also I have opened a Jira to rename chunkLayoutVersion -> ContainerLayOutVersion.

Adding 2 more versions is not a big change and it is simple, with one layout version code looks simple I believe. If you are strongly in favor of that, I can open a new Jira and for doing this.

hanishakoneru commented:
> But anyway I beleive our recommendation will be to use FILE_PER_BLOCK going forward. To support older layout version this is being done from my understanding how we have done in this PR.

In that case, why don't we go with FILE_PER_BLOCK approach and discard the FILE_PER_CHUNK going forward?

Also LayoutVersions should not be configurable. In my understanding, they should only be used when there is an incompatible change. Specifically with this change, this is not really a user configurable parameter. This version onwards metadata will always be stored in DB irrespective of what value "ozone.scm.chunk.layout" is configured to.

bharatviswa504 commented:
> In that case, why don't we go with FILE_PER_BLOCK approach and discard the FILE_PER_CHUNK going forward?

Yes, we can remove it if not needed. But I don't want to touch this in this Jira. I can open a new Jira and if we get consensus we can remove it.


`Specifically with this change, this is not really a user configurable parameter. This version onwards metadata will always be stored in DB irrespective of what value "ozone.scm.chunk.layout" is configured to.`

The reason for not removing the older version is for this, which I have mentioned in the code comments. Technically in newer software version I can remove FILE_PER_CHUNK and FILE_PER_BLOCK.

```
      // If on newer version of software if user configured mistakenly to use
      // older version, log an warning message. Not changing to newer layouts
      // for now to give an option to user if some bug present in newer layout
      // version, user has option to use still older version. We can change
      // this automatically in newer releases if no issues with newer version.
```

This is done for being on safer side.

hanishakoneru commented:
> // If on newer version of software if user configured mistakenly to use
      // older version, log an warning message. Not changing to newer layouts
      // for now to give an option to user if some bug present in newer layout
      // version, user has option to use still older version. We can change
      // this automatically in newer releases if no issues with newer version.

I don't think this is a good idea. If we add this option now, we might have to keep maintaining it forever and all future versioning might also need to be maintained in the same way.

bharatviswa504 commented:
> > // If on newer version of software if user configured mistakenly to use
> > // older version, log an warning message. Not changing to newer layouts
> > // for now to give an option to user if some bug present in newer layout
> > // version, user has option to use still older version. We can change
> > // this automatically in newer releases if no issues with newer version.
> 
> I don't think this is a good idea. If we add this option now, we might have to keep maintaining it forever and all future versioning might also need to be maintained in the same way.



> > // If on newer version of software if user configured mistakenly to use
> > // older version, log an warning message. Not changing to newer layouts
> > // for now to give an option to user if some bug present in newer layout
> > // version, user has option to use still older version. We can change
> > // this automatically in newer releases if no issues with newer version.
> 
> I don't think this is a good idea. If we add this option now, we might have to keep maintaining it forever and all future versioning might also need to be maintained in the same way.

Addressed the comment.

bharatviswa504 commented:
Thank You @adoroszlai for the offline discussion. I have removed adding a new layout version and updated the code to consider both old and new containers.

@hanishakoneru Updated the patch from last time, Have a look in to it.

bharatviswa504 commented on hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/keyvalue/helpers/KeyValueContainerUtil.java:
```diff
@@ -150,29 +160,109 @@ public static void parseKVContainerData(KeyValueContainerData kvContainerData,
+
+
+      boolean success = true;
+      while (success) {
```
This is done like this because hasNext will return StorageContainerException when unable to parse block data.

bharatviswa504 commented on hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/keyvalue/helpers/KeyValueContainerUtil.java:
```diff
@@ -150,29 +160,109 @@ public static void parseKVContainerData(KeyValueContainerData kvContainerData,
         } catch (IOException ex) {
-          return 0L;
+          LOG.error("Failed to parse block data for Container {}",
+              kvContainerData.getContainerID());
```
Previously also we used to do this. Slightly modified the code. And I think for one block corruption, if we don't add it to container set, then BackGround Container Scanner will never know about this container, and it will never be fixed. So, I think adding to container set is right thing to do here. Let me know your thoughts?

```
     bytesUsed = liveKeys.parallelStream().mapToLong(e-> {
        BlockData blockData;
        try {
          blockData = BlockUtils.getBlockData(e.getValue());
          return blockData.getSize();
        } catch (IOException ex) {
          return 0L;
        }
      }).sum();
```

hanishakoneru replied:
I see that previously initializeUsedBytes() was done only for OPEN containers. And for others, we would just get the BlockData and update usedBytes.  
What I mean to say is that the behavior has changed with this patch. I might have missed some background here. Can you please explain why the change.

bharatviswa504 replied:
Previously we used to compute twice, one using different logic and 2nd using different logic. Now we are doing only once by using initializeUsedBytesAndBlockCount logic.

bharatviswa504 replied:
And also at the end, even new logic which is using iterator finally uses BlockUtils.getBlockData to getNext Block. I feel the current way of iterating and computing the size is cleaner way, instead of getting all blocks into in-memory at once.

hanishakoneru replied:
Thanks for the explanation Bharat. Makes sense.

bharatviswa504 commented on hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/keyvalue/impl/BlockManagerImpl.java:
```diff
@@ -224,11 +239,20 @@ public void deleteBlock(Container container, BlockID blockID) throws
+      BatchOperation batch = new BatchOperation();
+      batch.delete(blockKey);
+      batch.put(DB_CONTAINER_BYTES_USED_KEY,
+          Longs.toByteArray(container.getContainerData().getBytesUsed()));
```
Right now deleteBlock in BlockManagerImpl is not used, BlockDeletingService delete Chunks and then delete the blocks which are marked for deleted from container DB.

And also updating bytes used is not correct here, so only updated block count here. As anyway bytes used is taken care during delete chunk.
```

        // Once files are deleted... replace deleting entries with deleted
        // entries
        BatchOperation batch = new BatchOperation();
        succeedBlocks.forEach(entry -> {
          String blockId =
              entry.substring(OzoneConsts.DELETING_KEY_PREFIX.length());
          String deletedEntry = OzoneConsts.DELETED_KEY_PREFIX + blockId;
          batch.put(DFSUtil.string2Bytes(deletedEntry),
              DFSUtil.string2Bytes(blockId));
          batch.delete(DFSUtil.string2Bytes(entry));
        });
```

bharatviswa504 commented on hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/common/statemachine/commandhandler/DeleteBlocksCommandHandler.java:
```diff
@@ -251,12 +253,23 @@ private void deleteKeyValueContainerBlocks(
+      // update pending deletion blocks count and delete transaction ID in
+      // in-memory container status
+      containerData.updateDeleteTransactionId(delTX.getTxID());
+
```
Good catch. Added code to update DB only when delete TxID > current Tx ID

bharatviswa504 commented:
Thank You @hanishakoneru for the review. 
Addressed review comments in the latest commit.

bharatviswa504 commented on hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/common/statemachine/commandhandler/DeleteBlocksCommandHandler.java:
```diff
@@ -256,9 +256,14 @@ private void deleteKeyValueContainerBlocks(
+      // In memory is updated only when existing delete transactionID is
+      // greater.
+      if (containerData.getDeleteTransactionId() > delTX.getTxID()) {
+        // Update in DB pending delete key count and delete transaction ID.
```
Done

bharatviswa504 commented:
> @bharat, the commits seem to have gone back to previous version. I see FILE_PER_BLOCK_AND_CONTAINER_DB_HAS_METADATA and versioning in the latest commits.

Thank You for review and finding this issue, I accidentally used my old branch and pushed it to the branch. Thanks for sharing my old patch I have addressed the review comment.

hanishakoneru commented:
+1 pending CI. 
Thanks @bharatviswa504 for working on this diligently.

bharatviswa504 commented:
Thank You @hanishakoneru for the review and @adoroszlai thanks for the initial review, if you have any more comments further happy to address them in separate Jira. (@adoroszlai is on vacation for 2 weeks, so committing for now as got one +1 from @hanishakoneru )
